lib/astutils.cpp

```diff
@@ -1212,45 +1212,46 @@ SmallVector<ReferenceToken> followAllReferences(const Token* tok,
             return x.token < y.token;
         }
     };
-    SmallVector<ReferenceToken> refs_result;
     if (!tok)
-        return refs_result;
+        return {};
     if (depth < 0) {
-        refs_result.push_back({tok, std::move(errors)});
```

Comment:
Is there a reason not to return an init list here (and below)? E,g, `return { ReferenceToken{tok, std::move(errors)}); };` or even `return { {tok, std::move(errors)} }`.

Reply:
There was some issue with that but I cannot remember what as those changes have been sitting so long in my tree already. Will look into it.

Reply:
I remember now - partially at least. `SmallVector` has no constructor with `std::initializer_list`. I had either issues with hooking up the allocator for that or it was slower.

Reply:
This seems to work code-wise:
~~~
    SmallVector(std::initializer_list<T> il)
        : std::vector<T, TaggedAllocator<T, N>>(il)
    {
        this->reserve(N);
    }
~~~

Reply:
Reserving the space after having already added the entries seems suspicious/unnecessary.

There also appears to be a bug with initializer lists containing only a single entry in Clang/MSVC: https://github.com/llvm/llvm-project/issues/72396.

Reply:
> Reserving the space after having already added the entries seems suspicious/unnecessary.

Yes, that's redundant.

> There also appears to be a bug with initializer lists containing only a single entry in Clang/MSVC: https://github.com/llvm/llvm-project/issues/72396.

The issue seems to be related to initialisation with the same type (`SmallVector`) though, not the element type (`T`),

Reply:
It compiles but it is slower since it now calls some allocation code with Boost instead of just moving that object out. `small_vector` is quite particular in how you have to use it.

Reply:
I'm getting this for 3x `push_back()` of `int`  vs equivalent init list:
~~~
BM_init_list       21.0 ns         21.0 ns     32000000
BM_push_back       4.39 ns         4.10 ns    179200000
~~~
Maybe this is worth reporting upstream?